### DISCUSSION TITLE
Library support for New Architecture + Bridgeless in 2024

### DISCUSSION THREAD

**Discussion by brentvatne** [reactwg/react-native-new-architecture#167] [Libraries]
> [!NOTE] 
> tl;dr: The New Architecture is being rolled out in 2024, and now is a great time to verify and fix compatibility issues in your libraries. React Native 0.74 provides an Interop Layer that greatly assists in this process, making it possible to use modules written for the legacy architecture with minimal changes in most cases. A small team of engineers have kicked off the effort with some of the most installed libraries, but more help is needed from the community to ensure coverage across the vast ecosystem. We're requesting help in the form of: library users testing and reporting issues, authors updating libraries, and engineers experienced within this domain assisting others with making their libraries compatible. Resources, links, and additional context are below.

> [!IMPORTANT]
> **The current status of many of the most popular libraries is tracked on [React Native Directory](https://reactnative.directory/?newArchitecture=true).**

As shared in [the New Architecture landing page](https://reactnative.dev/docs/next/the-new-architecture/landing-page), 2024 is the year that the New Architecture will be rolled out to all React Native users. 

In order for developers to be able to begin transitioning their existing projects to the New Architecture, it's important that the libraries they depend on are compatible.

There have been [previous efforts](https://github.com/reactwg/react-native-new-architecture/discussions/6) to expand library support for the New Architecture, which helped to inform changes that were needed to make this process easier for library maintainers. 

In React Native 0.74, there are various [**Interop Layers**](https://github.com/reactwg/react-native-new-architecture/discussions/135) enabled by default which allows many libraries to work without any changes — however, it's not perfect and some libraries will need to be updated to account for its quirks. 

React Native 0.74 also enables **bridgeless mode by default**, which is a key feature of the New Architecture and may require changes to some libraries.

This discussion is a call for support in migrating third party React Native libraries, in this final push to get the New Architecture shipped.

## Kicking off the initiative with the most frequently installed libraries

Meta and Expo kicked off this effort by identifying the **400 most installed libraries** on Expo's CI service ([EAS Build](https://docs.expo.dev/build/introduction/)), and then filtering those down to those that depended on native code. We believe this is a good indicator of which are the most popular libraries in the React Native ecosystem as a whole.

We proceeded to either work on testing and fixing the libraries ourselves, or reached out to the maintainers of those libraries to ask for support. Many of those folks kindly agreed to help, and we've been working together to get their libraries updated.

📈 🔗 **The current status of these libraries is tracked ~in the ["New Architecture support in popular libraries" spreadsheet](https://docs.google.com/spreadsheets/d/1F1tI9PQLl_uab3HNYNwgjPVL2r0hJ9VymXcgDp5etdg/edit?usp=sharing)~ on [React Native Directory](https://reactnative.directory/?newArchitecture=true).**

This list includes the most popular libraries in the ecosystem, including libraries from Software Mansion such as [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated/) and [react-native-screens](https://github.com/software-mansion/react-native-screens), libraries in the Expo SDK such as [expo-updates](https://docs.expo.dev/versions/latest/sdk/updates/), service libraries like [react-native-firebase](https://rnfirebase.io/), and many others.

As of Wednesday, June 26, 2024, >61% of the the 400 most installed libraries are compatible.

## How you can help

There are a few ways you can help with this effort, depending on your experience and interests:

- **Test libraries**: If you're a library user, you can test libraries that you depend on and report any issues you encounter. This will help library authors identify which libraries need to be updated. [Learn how to test your libraries](https://gist.github.com/cipolleschi/82b7a9561b8861330efabbd3eb08c6f5).
- **Update libraries**: If you're a library author/maintainer, or willing to contribute, you can update libraries to work with the New Architecture and bridgeless mode. Using the Interop Layer, this can often be relatively straightforward. You can also open a discussion in this repository to ask for help or provide updates on the status of your library. Read “[Enable the New Architecture for Libraries](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries.md)” to get started.
- **Help other library authors**: If you're familiar with the migration process, you can help other library authors update their libraries by answering their questions here or in issues on their repositories.

Be sure to refer to the [React Native Directory](https://reactnative.directory/?newArchitecture=true) to see which libraries are already supported. Additionally, you can search in this discussion to see if anyone has already opened a thread for the library you're interested in.

The libraries that are most likely to require updates are those that ship or depend on third party native code. Libraries that only use JavaScript or React Native core APIs are expected to work without any changes (provided that they do not use private/undocumented APIs).

### Resources to help you get started

- [Enable the New Architecture for Libraries](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries.md): the first document that you should read when you want to add New Architecture support for a library. This document was recently updated to recommend using the Interop Layer first for the shortest path to compatibility.
- [Test your library against React Native 0.74-rcs](https://gist.github.com/cipolleschi/82b7a9561b8861330efabbd3eb08c6f5): a straightforward guide to test your library against React Native 0.74-rcs.
- [Test your library built with the Expo Modules API against React Native 0.74-rcs](https://www.youtube.com/watch?v=B5ActXVussQ): a quick video walkthrough.
- [React Native New Architecture Working Group - Guides](https://github.com/reactwg/react-native-new-architecture): a collection of guides and resources to help developers understand the New Architecture.

### Discussion title

Use the library name, eg: react-native-safe-area-context.

### Discussion body

Provide information in the following format (where "New Architecture" means with bridgeless also enabled):

- **Library version**: The version of the library you tested.
- **React Native version**: The version of the React Native you tested with (eg: 0.74-rc.4).
- **Compatible with New Architecture?**: Yes/No/Unknown. If not, please provide details. Be sure to test both platforms (iOS and Android) and provide details for each.
- **Compatible with New Architecture disabled?**: Yes/No/Unknown. If not, please provide details. Be sure to test both platforms (iOS and Android) and provide details for each.

Below this, share any questions, comments, or concerns you have.

**Comment by mrousavy** [comment]
Btw - who was write access to that google doc with the libraries?

1. react-native-blurhash supports new arch (Fabric)
2. react-native-mmkv PR is ready, waiting on RN 0.74 release
3. react-native-vision-camera is still in progress (kinda blocked at the moment)

**Reply by cortinico** [reply]
> react-native-vision-camera is still in progress (kinda blocked at the moment)

What's your blocker here @mrousavy ?

**Reply by mrousavy** [reply]
VisionCamera is a bit tricky.

Most of VisionCamera is a "normal" Android (Java)/iOS (Swift) library. For example, the `<Camera>` component has props like:

- `audio={..}`
- `isActive={..}`
- `photoHdr={..}`
- `zoom={..}`

But then there is this one special prop called `frameProcessor={..}`, which is a `jsi::Function` that needs to be converted to a Frame Processor (which essentially is just a [`WKTJsiWorklet`](https://github.com/margelo/react-native-worklets-core/blob/8a4e971e4b42411e24979d8cc2ccd6d09ad7c804/cpp/WKTJsiWorklet.h#L81)) using JSI/C++.

In the old architecture that was a pain to achieve - since I had to get the `jsi::Runtime`, install a global function into the runtime called `global.__setFrameProcessor(...)` or something, get the `frameProcessor={..}` from props and [imperatively update it in `componentDidUpdate`](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/src/Camera.tsx#L447-L457), pass the Function and View Tag (`findNodeHandle(...)`) to that function, and then [inside `setFrameProcessor` I perform a View lookup, resolve & cast the view, and finally convert the `jsi::Function`  to a Worklet and set it on the Camera](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/ios/Frame%20Processor/VisionCameraProxy.mm#L68-L82).

While I could still build the system like it was before (with a global `__setFrameProcessor` function instead of just passing the prop to the Camera), I feel like this could be simplified a lot with TurboModules - so I was kinda hoping to get some thoughts from you guys about this.

I have multiple questions here:

1. Is it possible to create such type converters in C++, which basically defines a conversion from `jsi::Function` to `com.mrousavy.camera.FrameProcessor`/`FrameProcessor` Java/Swift objects? That'd be the ideal solution, since I can skip the entire getting Runtime access, global injection, view lookup, etc part.
2. If 1. is not possible, would it be possible to define one Java/Swift View Component, and then a C++ view manager additionally that can work with the Java/Swift View Component as a strong type (something like `std::shared_ptr<VisionCameraView>`), so I can route the `frameProcessor={..}` prop through the C++ view manager, and the rest through the default generated Java/Swift ones?
3. If both is not possible, can I somehow at least pass a reference to my C++ view to a TurboModule, so I can imperatively set that and avoid the `findNodeHandle` lookup?
4. I have a bunch of discriminating unions that are used as enums. For example:
    - [`resizeMode: 'cover' | 'contain'`](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/src/CameraProps.ts#L143-L150)
    - [`ResizeMode.kt`](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/android/src/main/java/com/mrousavy/camera/types/ResizeMode.kt#L5-L17) / [`ResizeMode.swift`](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/ios/Types/ResizeMode.swift#L12-L36)
    - Actual parsing in ViewManager: [Android](https://github.com/mrousavy/react-native-vision-camera/blob/83168044a66f18a5b36dc996a389ded7bb840a1c/package/android/src/main/java/com/mrousavy/camera/CameraViewManager.kt#L122-L130) 
    ...is that supported?
5. I have a function (`startRecording()`) that can take both callbacks (`onRecordingEnded` and `onRecordingError`), and return a Promise (resolved once recording actually started natively). With the old architecture/bridge, this was not possible. Is that possible now?
6. How does Swift support look like nowadays in Fabric/TurboModules/CodeGen? Do I need to [bridge through C++](https://www.swift.org/documentation/cxx-interop/status/) or Objective-C?

**Reply by brentvatne** [reply]
@mrousavy - I made you an editor! anyone who maintains one of the libraries there can ping me to get editor access

**Reply by brentvatne** [reply]
> react-native-blurhash supports new arch (Fabric)

did you test this against 0.74-rc with bridgeless enabled? [(how to do that)](https://gist.github.com/cipolleschi/82b7a9561b8861330efabbd3eb08c6f5)

**Reply by RSNara** [reply]
@sammy-SC, looping you in.

I think C++ TurboModules could at best clean up the __setFramProcessor binding. 

Is there any fabric infrastructure we could leverage to make this easier in the new architecture?

**Reply by mym0404** [reply]
@mrousavy 

I am developing [a package](https://github.com/mj-studio-library/react-native-naver-map) supports old/new arch both with codegen.

At a first glance, the codegen supporting is not comfortable yet.

Also I can swift but write it with objc because of lack of guide about that and sometimes we should access generated objc, c++ code.

Yes it seems we can use Swift(bridge header, c++ interop from5.9)but will be more tricky at now.

All you suggested sound like a experimental of experimental features at now...


---

btw can I ask why did you migrate from plain jsi to turbo/fabric?

I understand with plain JSI implementation, it works old,new arch both. Am I wrong?

**Reply by brentvatne** [reply]
@mym0404 - can you clarify who your question is directed at?

**Reply by mym0404** [reply]
sure

**Reply by mrousavy** [reply]
@RSNara @sammy-SC sorry hate to bump this, but do you have any thoughts on my questions? I'm not sure how to proceed with that in VisionCamera and your insights would be hugely appreciated - would be dope to use CodeGen/Fabric for this to avoid the whole global JSI function logic.

**Reply by mrousavy** [reply]
@brentvatne I haven't tested rn blurhash with bridgeless yet, but it doesn't use JSI so I think it should work fine. I'll double check :)

**Reply by mrousavy** [reply]
@mym0404 I am not migrating from JSI to Turbo/Fabric, I am migrating from old arch to new arch. There are some parts that I wrote in JSI which might be possible to do with new arch (Turbo/Fabric/CodeGen), so that would be less code for me to write, a bit safer since I no longer dangerously look up the view using global functions, and type-safe, but I'm waiting on the Meta team to share some thoughts on this if this is possible to do. :)

**Reply by mrousavy** [reply]
@philIip asked me on Discord to clarify my question a bit, so here we go:


I think the VisionCameraExample app (inside the repo, `package/example`) is a perfect way to demonstrate this.

You can follow along the route it goes quite easily both in JS and native:

1. In JS, the user uses the `useFrameProcessor` hook (which essentially just returns the same function you passed into it, but with a custom try/catch error handler) and passes the `frameProcessor` function to the `<Camera>` **just like any other prop**:
    ![image](https://github.com/reactwg/react-native-new-architecture/assets/15199031/e7b03c71-83d4-493f-916c-1f4594c52f57)

2. If we just pass the `frameProcessor` prop to the native `<NativeCameraView>`, it will be converted to a `RCTDirectEventBlock`, which we don't want. We need to use JSI because we send the Frames as custom `jsi::HostObjects` through to JS. So we cannot use the Native Module type conversion system here.
  Instead, we use `componentDidUpdate()` to **imperatively** set the frameProcessor on the view using JSI functions:
    ![image-2](https://github.com/reactwg/react-native-new-architecture/assets/15199031/f9c9e7aa-49d1-41b6-8d20-c2034b889ef1)

3. Inside the imperative `setFrameProcessor(...)` (or `unsetFrameProcessor()`) functions, we call a global JSI method that I injected into the Runtime, this takes a **view tag** (`findNodeHandle`) and the frame processor itself:
    ![image-3](https://github.com/reactwg/react-native-new-architecture/assets/15199031/62bd5830-758e-471e-8007-2c5b7a9ca867)

4. And from then on we are now already in native (Objective-C++ or C++/JNI) land.
    1. I do some conversions on the `jsi::Function` (`arguments[1]`) to make it a Worklet which is callable from another Thread (see line 69-70)
    2. I resolve the `CameraView*` using it's viewTag (`arguments[0]`) (this currently happens through `[UIManager viewForReactTag:]` on iOS) (see line 73-75)
    3. I create an instance of a custom type I created (called `FrameProcessor`, it just holds the C++ Worklet instance and exposes it to ObjC/Java), and set it on the `CameraView`.
    ![image-4](https://github.com/reactwg/react-native-new-architecture/assets/15199031/00b53f12-e172-402b-b0da-4727c70dc5f2)

5. This is the `FrameProcessor` type that I set on the view (same exists for Java/C++/JNI).
    ![Screenshot_2024-04-09_at_11 10 45](https://github.com/reactwg/react-native-new-architecture/assets/15199031/7a685edf-12d4-4925-9927-7064f6876adf)

6. As you can see, it's treated like a normal prop from React - just like all other props. It's just the complex route it has to go through for the conversion from `jsi::Function` -> `Worklet` -> `FrameProcessor` that makes this hard to achieve.
    ![image-5](https://github.com/reactwg/react-native-new-architecture/assets/15199031/c0d54c8b-2ede-412a-ad8f-9695c7375580)

Essentially what I am looking for is to expose the `frameProcessor` property to React just like all the other props that you can see in 6., and have my custom C++ type converter that creates the `FrameProcessor` object from a `jsi::Function`. Ideally I just want to write that part, just like an `RCT_ENUM_CONVERTER`, but for the jsi::Function now.

Is that possible?

**Reply by angelo-hub** [reply]
following up here as i'm building birdgeless right now and in the next 2 months will be implementing `react-native-vision`

**Comment by TheRogue76** [comment]
@brentvatne I sent you an email regarding getting access to the doc. Hope that's ok. Just need to add the details for lib version and the versions it was tested against

**Reply by brentvatne** [reply]
sure thing! I've sent you an invite to edit it

**Comment by leonelmq** [comment]
Hi!  I found firebase SDK already in the list.  Is it planned to reach adobe to add the whole SDK to the migration?
https://github.com/adobe/aepsdk-react-native

**Reply by brentvatne** [reply]
hey @leonelmq! if you use this adobe library, it would be great for you to reach out through their issues and request this feature. you can point them to this discussion for context

**Comment by axeldelafosse** [comment]
Hey Brent! Thank you so much for this initiative, this is great!

Can we add `zeego` and its dependencies please? It's a great lib by @nandorojo :)

Here is the list:
- [`zeego`](https://www.npmjs.com/package/zeego)
- [`@react-native-menu/menu`](https://www.npmjs.com/package/@react-native-menu/menu)
- [`react-native-ios-context-menu`](https://www.npmjs.com/package/react-native-ios-context-menu)
- [`react-native-ios-utilities`](https://www.npmjs.com/package/react-native-ios-utilities)

**Reply by nandorojo** [reply]
Brent actually tested this on iOS last I spoke to him, and while there's no crash, it doesn't work. You can see the issue I opened at https://github.com/dominicstop/react-native-ios-context-menu/issues/100

**Reply by brentvatne** [reply]
done! I added them to the "Community reported" section of the sheet

**Comment by SerSerch** [comment]


**Reply by brentvatne** [reply]
hi @SerSerch! re: expo support - any module that works without the expo module system set up should work with it set up. they are typically orthogonal. the only exception is when the module requires you to use its own subclass of eg: appdelegate

**Reply by SerSerch** [reply]
@brentvatne oh, thanks! it works!
[expo-app](https://github.com/serserm/react-native-turbo-serialport/tree/main/examples/expo-app)

**Comment by younes0** [comment]
@brentvatne 
Could you add please: 
- react-native-fast-shadow:⚠️ not implemented, see https://github.com/alan-eu/react-native-fast-shadow/issues/25
- react-native-ux-cam: ⚠️ not implemented, see https://github.com/uxcam/react-native-ux-cam/issues/132
- react-native-simple-toast: ✔️complete: see https://github.com/vonovak/react-native-simple-toast

And update eventually:
- @sentry/react-native: ✔️complete: see: https://docs.sentry.io/platforms/react-native/migration/#from-4x-to-5x

**Reply by brentvatne** [reply]
done! I added it to the "Community reported" section of the sheet

**Comment by ngocle2497** [comment]
@brentvatne 
Can you add:
- okta-react-native: created [PR](https://github.com/okta/okta-react-native/pull/422)

**Reply by brentvatne** [reply]
done! I added it to the "Community reported" section of the sheet

**Comment by TheLonelyAstronaut** [comment]
Hey @cortinico!

I'n currently porting React Native Navigation by Wix to new bridgeless architecture and have some questions/blockers:

1. What should i use instead of bridge.uiManager in case of bridgeless mode? I want to find view by nativeID, but C++ UIManager from RCTHost does not provide such functionality.
2. How could i get NSTextStorage from RCTParagraphComponentView? I need it for shared element transition.

**Comment by ngocle2497** [comment]
Hi @cortinico , just create [PR](https://github.com/wonday/react-native-orientation-locker/pull/297) to migratte react-native-orientation-locker to new arch. Can u update the sheet?

**Reply by brentvatne** [reply]
done! I added it to the "Community reported" section of the sheet

**Comment by morenoh149** [comment]
hi @cortinico I have a [PR](https://github.com/morenoh149/react-native-contacts/pull/739) to migrate react-native-contacts to new arch. Can you update the sheet?

**Reply by cortinico** [reply]
cc @brentvatne

**Reply by brentvatne** [reply]
@morenoh149 - you can update the entry for it on react-native-directory, this is where the canonical compatibility information is stored now: https://github.com/react-native-community/directory

**Comment by 0xFA11** [comment]
I think we should also include @gorhom's [react-native-bottom-sheet](https://github.com/gorhom/react-native-bottom-sheet) in the list.

2 useful links:
- https://github.com/gorhom/react-native-bottom-sheet/issues/1726
- https://github.com/gorhom/react-native-bottom-sheet/pull/1823

Hopefully, `react-native-bottom-sheet` v5 could implement and support new arch!

**Reply by brentvatne** [reply]
@ThusSpokeNomad - you can update the entry for it on react-native-directory, this is where the canonical compatibility information is stored now: https://github.com/react-native-community/directory

**Reply by 0xFA11** [reply]
@brentvatne done, see here: https://github.com/react-native-community/directory/pull/1198

**Comment by siddarthkay** [comment]
It would be nice to include Tesla Motors `react-native-camera-kit` -> https://github.com/teslamotors/react-native-camera-kit

I tried to enable new architecture for our app and observed that this library works well on iOS but shows this warning on Android.
![Screenshot 2024-06-05 at 12 35 17 PM](https://github.com/reactwg/react-native-new-architecture/assets/64726664/a0b4eba5-2971-47ca-b94f-bfdc407ad16f)

The library maintainers will welcome PRs regarding new architecture -> https://github.com/teslamotors/react-native-camera-kit/issues/537#issuecomment-1613619321

Here is my research issue where I document our journey to enabling new architecture for status-mobile -> https://github.com/status-im/status-mobile/issues/18138

**Reply by brentvatne** [reply]
@siddarthkay - you can update the entry for it on react-native-directory, this is where the canonical compatibility information is stored now: https://github.com/react-native-community/directory

**Reply by siddarthkay** [reply]
@brentvatne : https://github.com/react-native-community/directory/pull/1289

**Comment by vipindvipin** [comment]
Hi @brentvatne  @cortinico

I've recently migrated my React Native project to the New Architecture (currently using **React Native version 0.72.14**) and will be maintaining this setup for a few months. 

After the migration, multiple issues have surfaced. For example, components like FastImageView and BVLinearGradient stopped working. To address this, I opted for the interop layer by adding these components in react-native-config under unstable_reactLegacyComponentNames, and they started working.

As I'm not updating the React Native version, I’m left with using the interop approach**. However, I encounter challenges in finding the correct component names to add to react-native-config.**

The bigger concern now is that my app has become very slow, with unresponsive touches and sluggish bottom navigation. The performance issues seem to be related to the incompatibility of libraries with the new architecture.

**How can we efficiently identify the correct component names to add to react-native-config for interop. Determining which specific names are needed for each library is often confusing.**

What are the other possible ways to tackle major issues like unresponsive touches and slow bottom navigation?

**Note: Upgrading the React Native version is not an option for me.**

**Reply by cortinico** [reply]
Hey @vipindvipin 

> How can we efficiently identify the correct component names to add to react-native-config for interop. Determining which specific names are needed for each library is often confusing.

Since React Native 0.74, there is no need to specify the components you want to add to the interop layer: https://github.com/reactwg/react-native-new-architecture/discussions/175

> Note: Upgrading the React Native version is not an option for me.

Sadly if you don't upgrade React Native version you will have a way "worse" New Architecture experience. We shipped several bugfixes in the last months that you'll be missing out. Moreover, you won't be able to distribute to the PlayStore anymore once the SDK requirements by Google are enforced, so I invite you to consider upgrading to the latest stable.

**Comment by gunnartorfis** [comment]
###  @adyen/react-native

Library version: 2.4.0
React Native version: 0.74.5
Compatible with New Architecture? No

**Crashes on iOS**, verified by removing every import from @adyen/react-native. 

**Build fails on Android** with a "Java heap space" error. 
<img width="1888" alt="Screenshot 2024-09-16 at 08 46 36" src="https://github.com/user-attachments/assets/1813ddc0-c966-4ea2-9cb6-15bb8b1042f6">

**Reply by cortinico** [reply]
> Build fails on Android with a "Java heap space" error.

That's unrelated to the New Architecture by the way. It's a Gradle/Android issue that happens when Gradle hasn't enough memory to execute.
